clippy_lints/src/lib.rs

```diff
@@ -162,6 +162,8 @@ macro_rules! extract_msrv_attr {
 mod consts;
 #[macro_use]
 mod utils;
+#[cfg(feature = "metadata-collector-lint")]
+mod deprecated_lints;
```

Comment:
Wait, this module was dead the whole time? :open_mouth: In that case: couldn't you collect the necessary metadata from the `register_removed` calls, so we could remove this module once we completely switch over to this metadata collector?

Reply:
Yes it was only used by the collection script ^^.

Well, I decided to do it this way because [`deprecated_lints.rs`](https://github.com/rust-lang/rust-clippy/blob/master/clippy_lints/src/deprecated_lints.rs) contains a more in-depth explanation why these lints have been depreciated in comparison to our usage of [store.register_removed](https://github.com/rust-lang/rust-clippy/blob/65951c969ff843bc3a45c470d5f366498172ebe1/clippy_lints/src/lib.rs#L444). But that could easily be updated.

However, I think there is currently no way to retrieve the deprecation reason from the `LintStore`. Removed lints are only stored in the [`by_name`](https://doc.rust-lang.org/stable/nightly-rustc/rustc_lint/struct.LintStore.html#structfield.by_name) field which can not be accessed directly. The function [`find_lints`](https://doc.rust-lang.org/stable/nightly-rustc/rustc_lint/struct.LintStore.html#method.find_lints) can be used to access the `by_name` field but will only return `Err(FindLintError::Removed)` without the reason itself (See [rustc_lint/context.rs](https://doc.rust-lang.org/stable/nightly-rustc/src/rustc_lint/context.rs.html#314-333)). I could create an issue or PR in [rust-lang/rust](https://github.com/rust-lang/rust) to maybe add a new function to the `LintStore` for this. That might also be cleaner when I think about it, but it would probably take some time until that would be implemented and synced back. I could add it to #7172 as a new to-do item. What is your opinion on the matter?

Reply:
Yes, the implementation uses the `LintStore` to collect the lint group, as it was the easiest way to retrieve it. ([here](https://github.com/rust-lang/rust-clippy/blob/72be4764c57266dc1e71d32a06619d302b3253a4/clippy_lints/src/utils/internal_lints/metadata_collector.rs#L337))

Checking for `register_removed` would also work but checking each expression seemed like an unnecessary overhead since they all get registered in the store already anyways :thinking:

Reply:
> Yes, the implementation uses the `LintStore` to collect the lint group, as it was the easiest way to retrieve it.

Ah, ok. In that case my suggestion to only register this lint, when collecting metadata won't work.

---

But IIUC you're not using the `LintStore` for the deprecated lints anyway? So I don't think it would make a difference if you do this in `check_expr` or in `check_item` and match on the method call instead of the static item.

Reply:
> Ah, ok. In that case my suggestion to only register this lint, when collecting metadata won't work.

There are other ways to collect the lint group so it's not impossible. And your suggestion was focused on performance, we can simply skip the registration of the actual lint pass (`register_early_pass` and `register_late_pass`). Correct?

---

> So I don't think it would make a difference if you do this in `check_expr` or in `check_item` and match on the method call instead of the static item. 

Correct, I take it that you would prefer the expression via `check_expr`?

Reply:
> we can simply skip the registration of the actual lint pass (`register_early_pass` and `register_late_pass`).

Yeah, right. I always forget that registering lints and registering passes is done separately. 



> Correct, I take it that you would prefer the expression via `check_expr`?

Since we can get rid of a whole dead module that way (not immediately, but in the near future), yes definitely.